src/shims/x86/sse2.rs

```diff
@@ -0,0 +1,1042 @@
+
+                    this.write_scalar(Scalar::from_u64(res.into()), &dest)?;
+                }
+            }
+            // Used to implement the _mm_{sll,srl,sra}_epi16 functions.
+            // Shifts 16-bit packed integers in left by the amount in right.
+            // Both operands are vectors of 16-bit integers. However, right is
+            // interpreted as a single 64-bit integer (remaining bits are ignored).
+            // For logic shifts, when right is larger than 15, zero is produced.
+            // For arithmetic shifts, when right is larger than 15, the sign bit
+            // is copied to remaining bits.
+            "psll.w" | "psrl.w" | "psra.w" => {
+                let [left, right] =
+                    this.check_shim(abi, Abi::C { unwind: false }, link_name, args)?;
+
+                let (left, left_len) = this.operand_to_simd(left)?;
+                let (right, right_len) = this.operand_to_simd(right)?;
+                let (dest, dest_len) = this.place_to_simd(dest)?;
+
+                assert_eq!(dest_len, left_len);
+                assert_eq!(dest_len, right_len);
+
+                enum ShiftOp {
+                    Sll,
+                    Srl,
+                    Sra,
+                }
+                let which = match unprefixed_name {
+                    "psll.w" => ShiftOp::Sll,
+                    "psrl.w" => ShiftOp::Srl,
+                    "psra.w" => ShiftOp::Sra,
+                    _ => unreachable!(),
+                };
+
+                // Build the 64-bit shift operand from 16-bit elements
+                let shift = u64::from(this.read_scalar(&this.project_index(&right, 0)?)?.to_u16()?)
+                    | (u64::from(this.read_scalar(&this.project_index(&right, 1)?)?.to_u16()?)
+                        << 16)
+                    | (u64::from(this.read_scalar(&this.project_index(&right, 2)?)?.to_u16()?)
+                        << 32)
+                    | (u64::from(this.read_scalar(&this.project_index(&right, 3)?)?.to_u16()?)
+                        << 48);
+
+                for i in 0..dest_len {
+                    let left = this.read_scalar(&this.project_index(&left, i)?)?.to_u16()?;
+                    let dest = this.project_index(&dest, i)?;
+
+                    let res = match which {
+                        ShiftOp::Sll =>
+                            left.checked_shl(shift.try_into().unwrap_or(u32::MAX)).unwrap_or(0),
+                        ShiftOp::Srl =>
+                            left.checked_shr(shift.try_into().unwrap_or(u32::MAX)).unwrap_or(0),
+                        #[allow(clippy::cast_possible_wrap, clippy::cast_sign_loss)]
+                        ShiftOp::Sra => {
+                            // Convert u16 to i16 to use arithmetic shift
+                            let left = left as i16;
+                            left.checked_shr(shift.try_into().unwrap_or(u32::MAX))
+                                .unwrap_or(left >> 15) as u16
```

Comment:
What the `left >> 15` doing here? Is that the "copying the sign bit to the remaining bits"? Please add a comment, such random integer constants are always suspicious.